{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "microbiome_path = \"/home/kunal/tscc_projects/tanner/data/family3/microbiome/14009b/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of files and directories in this folder\r\n",
      "\r\n",
      "README.txt: \tthis file\r\n",
      "\r\n",
      "Samples.txt: \ta text file with a list of sample names\r\n",
      "        For projects that we received samples by batches, and we processed and annotated these samples\r\n",
      "        by batches, there will be additional files <Sample-batch-number.txt>.\r\n",
      "\r\n",
      "Sample directories: \teach sample directory has a list of files:\r\n",
      "        taxonomy-data: \t\ttaxonomy binning results for all genomes in HLI's reference genome database\t\r\n",
      "        superkingdom-reads-count.txt:\tnumber of reads mapped to different superkingdoms\r\n",
      "\r\n",
      "        assembles.fna:\t\tassemble sequences in FASTA format\r\n",
      "        assembles-cov:\t\taverage coverage of assembles\r\n",
      "        species-bins-assembles:\tassembles binned at species level\r\n",
      "        ORF.faa:\t\tORF sequences predicted from assembles in FASTA format\r\n",
      "        ORF-cov:\t\taverage coverage of ORFs\r\n",
      "        species-bins-ORFs:\tORFs binned at species level\r\n",
      "        ORF-single.faa:\t\tORF sequences predicted from un-assembled reads in FASTA format\r\n",
      "\r\n",
      "        Bacteria-phylum-abundance.txt:\trelative abundance at phylum level within Bacteria and Archaea species\r\n",
      "        Bacteria-class-abundance.txt:\tat class level\r\n",
      "        Bacteria-order-abundance.txt:\tat order level\r\n",
      "        Bacteria-family-abundance.txt:\tat family level\r\n",
      "        Bacteria-genus-abundance.txt:\tat genus level\r\n",
      "        Bacteria-species-abundance.txt: at species level\r\n",
      "        Bacteria-toprank-abundance.txt: at most top taxonomy rank\r\n",
      "\r\n",
      "        Eukaryota-phylum-abundance.txt:\trelative abundance within Eukaryota species\r\n",
      "        Eukaryota-class-abundance.txt\r\n",
      "        Eukaryota-order-abundance.txt\r\n",
      "        Eukaryota-family-abundance.txt\r\n",
      "        Eukaryota-genus-abundance.txt\r\n",
      "        Eukaryota-species-abundance.txt\r\n",
      "        Eukaryota-toprank-abundance.txt\r\n",
      "\r\n",
      "        Viruses-phylum-abundance.txt:\trelative abundance within viruses\r\n",
      "        Viruses-class-abundance.txt\r\n",
      "        Viruses-order-abundance.txt\r\n",
      "        Viruses-family-abundance.txt\r\n",
      "        Viruses-genus-abundance.txt\r\n",
      "        Viruses-species-abundance.txt\r\n",
      "        Viruses-toprank-abundance.txt\r\n",
      "\r\n",
      "        cog-ann.txt:\t\t\tCOG annotation at family level\r\n",
      "        cog-class-ann.txt:\t\tCOG annotation at class level\r\n",
      "        kog-ann.txt:\t\t\tKOG annotation at family level\r\n",
      "        kog-class-ann.txt:\t\tKOG annotation at class level\r\n",
      "        pfam-ann.txt:\t\t\tPfam annotation at domain/family level\r\n",
      "        tigrfam-ann.txt:\t\tTigrfam annotation at domain/family level\r\n",
      "        tigrfam-subrole-ann.txt:\tTigrfam annotation at subrole level\r\n",
      "        tigrfam-mainrole-ann.txt:\tTigrfam annotation at mainrole level\r\n",
      "        protein-ann.txt:\t\tProteins found in HLI reference protein database\r\n",
      "\r\n",
      "        More information of file format for taxonomy and function annotation is available in following section:\r\n",
      "\r\n",
      "        pathways directory:\t\tThis directory has pathway data for the species bins. Each species bin is either a collection of contigs that have the same species assignment or it consists of a single contig (>100 Kbp) that does not have a species level taxonomic assignment. For each species bin, there are following files:\r\n",
      "\t\tpathways-bin-speces-taxid.txt: \tpathway report generated by Pathway tools\r\n",
      "\t\tPathways-bin-speces-taxid:\tpathways found\r\n",
      "\t\tCompounds-bin-speces-taxid:\tcompounds found\r\n",
      "\t\tOperons-bin-speces-taxid:\toperons found\r\n",
      "\t\tProteins-bin-speces-taxid:\tproteins found\r\n",
      "\t\tReactions-bin-speces-taxid:\treactions found\r\n",
      "\t\tTC-bin-speces-taxid:\t\ttransport chemicals found\r\n",
      "\t\tTransporters-bin-speces-taxid:\ttransporters found\r\n",
      "\r\n",
      "\tMore information for pathway annotations is available at the end of this README file for following files\r\n",
      "\t\tCompounds-orgs-cnts:\t\tCompounds_matrix:   \r\n",
      "\t\tPathways-orgs-cnts:\t\tPathways_matrix\r\n",
      "\t\tProteins-orgs-cnts:\t\tProteins_matrix\r\n",
      "\t\tReactions-orgs-cnts:\t\tReactions_matrix\r\n",
      "\t\tTAM-orgs-cnts:\t\t\tTAM_matrix\r\n",
      "\t\tTC-orgs-cnts:\t\t\tTC_matrix\r\n",
      "\t\tTransporters-orgs-cnts:\t\tTransporters_matrix\r\n",
      "\r\n",
      "==================== More information of taxonomy and function annotation ====================\r\n",
      "\r\n",
      "In files cog-ann.txt, kog-ann.txt, pfam-ann.txt tigrfam-ann.txt and protein-ann.txt\r\n",
      "\tNo_refs: \tnumber of ORFs hit the reference protein or protein family\r\n",
      "\tCoverage:\tcoverage of this reference family, which is total aligned length multipled by ORF depth\r\n",
      "\t\t\tfrom all the ORFs that hit this reference divided by the length of the reference protein\r\n",
      "\tAbundance:\trelative abundance across all referende protein\r\n",
      "\r\n",
      "In files cog-class-ann.txt, kog-class-ann.txt, tigrfam-subrole-ann.txt tigrfam-mainrole-ann.txt \r\n",
      "\tNo_families:\tnumber of protein families (e.g. cog families) in this class\r\n",
      "\tCoverage:\tnumber of protein families found in this sample divided by number of protein families\r\n",
      "\tAbundance:\tsum of abundance from each protein families found in this sample\r\n",
      "\r\n",
      "In file species-bins-assembles\r\n",
      "\tA new bin starts with one line \">>>> taxid tax_name genome_length num_contigs len_contigs genome_cov\" and followed by\r\n",
      "\ta list of contig names in this bin\r\n",
      "\t\r\n",
      "In file species-bins-ORFs\r\n",
      "\tA new bin starts with one line \">>>> taxid tax_name num_ORFS\" and followed by \r\n",
      "        a list of ORF names in this bin\r\n",
      "\r\n",
      "\r\n",
      "==================== More information of pathway annotation ====================\r\n",
      "\r\n",
      "In Pathway annotations, classes \"describe general types of biological objects, \r\n",
      "such as the class of all genes\". Each file with one of the following names in \r\n",
      "the pathways directories contains a list of class instances from the MetaCyc pgdb \r\n",
      "(Pathways/Genome Database). The taxonomic distribution of MetaCyc can be \r\n",
      "found at http://metacyc.org/META/meta-org-pwy-stats. \r\n",
      "\r\n",
      "Compounds\r\n",
      "Operons\r\n",
      "Pathways \r\n",
      "Proteins\r\n",
      "Reactions\r\n",
      "TAM\r\n",
      "TC\r\n",
      "Transporters\r\n",
      "\r\n",
      "Each class is described as follows in the Pathway Tools version 18.5 User's Guide:\r\n",
      "\r\n",
      "Compounds\r\n",
      "--------- \r\n",
      "The Class Compounds describe small-molecular-weight chemical compounds typically, \r\n",
      "compounds that are substrates of metabolic reactions or compounds that activate or \r\n",
      "inhibit metabolic enzymes.\r\n",
      "\r\n",
      "Operons\r\n",
      "-------\r\n",
      "Operons refer to multiple-gene transcription units only, whereas 'transcription \r\n",
      "units' or 'TUs' include both operons and single-gene transcription units.\r\n",
      "\r\n",
      "Pathways\r\n",
      "--------\r\n",
      "Frames in class Pathways encode metabolic and signaling pathways.\r\n",
      "\r\n",
      "Proteins\r\n",
      "--------\r\n",
      "The class of all proteins is divided into two subclasses: protein complexes and polypeptides. \r\n",
      "A polypeptide is a single amino acid chain produced from a single gene. A protein complex \r\n",
      "is a multimeric aggregation of more than one polypeptide subunit. A protein complex may in \r\n",
      "some cases have another protein complex as a component. Many of the slots that are applicable \r\n",
      "to Proteins are also applicable to members of the RNA class.\r\n",
      "\r\n",
      "Reactions\r\n",
      "---------\r\n",
      "Frames within the Reactions class describe properties of a biochemical reaction independent \r\n",
      "of any enzyme or enzymes that catalyze that reaction. A reaction is a biochemical transformation \r\n",
      "that interconverts two sets of chemical compounds (which includes small metabolites, proteins, \r\n",
      "and DNA regions), and may translocate compounds from one cellular compartment to another. \r\n",
      "Most reactions are written in a conventional direction that has been assigned by the Enzyme Nomencla \r\n",
      "ture Commission, but that direction may or may not be the predominate physiological direction of \r\n",
      "the reaction. Reaction substrates can include small-molecular-weight compounds \r\n",
      "(for metabolic reactions), proteins (such as in signaling pathways), and DNA sites \r\n",
      "(such as for reactions involving binding of transcription factors to DNA).\r\n",
      "\r\n",
      "TAM \r\n",
      "---\r\n",
      "\"Transport Across Membrane\" is the output of the all_transport_across function from the pythoncyc api. \r\n",
      "The function returns a list of transport proteins that transport across one of the given membranes, \r\n",
      "which defaults to CCO-MEMBRANES defined by the Cellular Component Ontology, \r\n",
      "http://geneontology.org/page/cellular-component-ontology-guidelines.\r\n",
      "\r\n",
      "TC\r\n",
      "---\r\n",
      "\"Transport Chemicals\" is the output of the all_transport_chemicals function from the pythoncyc api. \r\n",
      "The function enumerates all chemicals transported by transport reactions in a species bin.\r\n",
      "\r\n",
      "Transporters\r\n",
      "------------\r\n",
      "Proteins predicted by Pathway Tools' Transport Inference Parser (TIP), which identifies proteins \r\n",
      "in a species bin that are likely to catalyze transport reactions.\r\n",
      "\r\n",
      "\r\n",
      "Files and File Formats\r\n",
      "----------------------\r\n",
      "The <class> placeholder represents a class as previously described. A file called Compounds, \r\n",
      "for example, lists all instances of the class compound from the pgdb of a species bin. \r\n",
      "Files and file formats are described below. File columns are tab delimited.\r\n",
      "\r\n",
      "<class>: List of class instances in MetaCyc.    \r\n",
      "format:    frame-id    common_name\r\n",
      "\r\n",
      "<class>-<species bin number>-<species>-<taxid>:    Class instances found for each species bin.\r\n",
      "format:    frame-id    common_name\r\n",
      "\r\n",
      "<class>-orgs-cnts:    Counts of class instances found across all species bins.\r\n",
      "format:    count    frame-id    common_name\r\n",
      "\r\n",
      "<class>_matrix:    A matrix of the number of times each MetaCyc class instance was found in each species bin. \r\n",
      "Column N+1 of the matrix corresponds to the Nth row in the <class> file. If a <class>-meta_cnts file is empty, \r\n",
      "then the <class>_matrix file will be a matrix of zeros. This matrix is not available for the Operons class. \r\n",
      "\r\n",
      "format: species_bin1    count_of_compound1    count_of_compound2    count_of_compound3    ... etc\r\n",
      "        species_bin2    count_of_compound1    count_of_compound2    count_of_compound3    ... etc\r\n",
      "        species_bin3    count_of_compound1    count_of_compound2    count_of_compound3    ... etc\r\n",
      "            .                    .                    .                    .            \r\n",
      "            .                    .                    .                    .            \r\n",
      "            .                    .                    .                    .            \r\n",
      "           etc                  etc                  etc                  etc\r\n",
      "\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "!cat $microbiome_path/README.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "manifest = \"/home/kunal/tscc_projects/tanner/doc/Nof1samples_manifest.xlsx\"\n",
    "manifest = pd.read_excel(manifest, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FamilyID</th>\n",
       "      <th>Individual ID</th>\n",
       "      <th>Date</th>\n",
       "      <th>Type</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>VendorID</th>\n",
       "      <th>Individual ID 2</th>\n",
       "      <th>Sample ID 1</th>\n",
       "      <th>Sample ID 2</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Individual Key</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "      <th>Family Key</th>\n",
       "      <th>Unnamed: 13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td> 1</td>\n",
       "      <td> CaLo</td>\n",
       "      <td> 04-16-14</td>\n",
       "      <td> Metabolites</td>\n",
       "      <td> Plasma</td>\n",
       "      <td> JCVI-00001</td>\n",
       "      <td> 2</td>\n",
       "      <td> 001_CaLo_04-16-14_Metabolites_JCVI-00001</td>\n",
       "      <td> 001_002_Metabolites_Plasma_JCVI-00001_04-16-14</td>\n",
       "      <td>NaN</td>\n",
       "      <td> TaLo</td>\n",
       "      <td> 1</td>\n",
       "      <td> Longstreet</td>\n",
       "      <td>  1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td> 1</td>\n",
       "      <td> CaLo</td>\n",
       "      <td> 04-16-14</td>\n",
       "      <td> Metabolites</td>\n",
       "      <td> Plasma</td>\n",
       "      <td> JCVI-00002</td>\n",
       "      <td> 2</td>\n",
       "      <td> 001_CaLo_04-16-14_Metabolites_JCVI-00002</td>\n",
       "      <td> 001_002_Metabolites_Plasma_JCVI-00002_04-16-14</td>\n",
       "      <td>NaN</td>\n",
       "      <td> CaLo</td>\n",
       "      <td> 2</td>\n",
       "      <td>        NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td> 1</td>\n",
       "      <td> CaLo</td>\n",
       "      <td> 07-11-14</td>\n",
       "      <td> Metabolites</td>\n",
       "      <td> Plasma</td>\n",
       "      <td> JCVI-00003</td>\n",
       "      <td> 2</td>\n",
       "      <td> 001_CaLo_07-11-14_Metabolites_JCVI-00003</td>\n",
       "      <td> 001_002_Metabolites_Plasma_JCVI-00003_07-11-14</td>\n",
       "      <td>NaN</td>\n",
       "      <td> MaLo</td>\n",
       "      <td> 3</td>\n",
       "      <td>        NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td> 1</td>\n",
       "      <td> CaLo</td>\n",
       "      <td> 09-25-14</td>\n",
       "      <td> Metabolites</td>\n",
       "      <td> Plasma</td>\n",
       "      <td> JCVI-00008</td>\n",
       "      <td> 2</td>\n",
       "      <td> 001_CaLo_09-25-14_Metabolites_JCVI-00008</td>\n",
       "      <td> 001_002_Metabolites_Plasma_JCVI-00008_09-25-14</td>\n",
       "      <td>NaN</td>\n",
       "      <td> GrLo</td>\n",
       "      <td> 4</td>\n",
       "      <td>        NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td> 1</td>\n",
       "      <td> MaLo</td>\n",
       "      <td> 09-25-14</td>\n",
       "      <td> Metabolites</td>\n",
       "      <td> Plasma</td>\n",
       "      <td> JCVI-00009</td>\n",
       "      <td> 3</td>\n",
       "      <td> 001_MaLo_09-25-14_Metabolites_JCVI-00009</td>\n",
       "      <td> 001_003_Metabolites_Plasma_JCVI-00009_09-25-14</td>\n",
       "      <td>NaN</td>\n",
       "      <td> ElGo</td>\n",
       "      <td> 5</td>\n",
       "      <td>        NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   FamilyID Individual ID      Date         Type  Tissue    VendorID  \\\n",
       "0         1          CaLo  04-16-14  Metabolites  Plasma  JCVI-00001   \n",
       "1         1          CaLo  04-16-14  Metabolites  Plasma  JCVI-00002   \n",
       "2         1          CaLo  07-11-14  Metabolites  Plasma  JCVI-00003   \n",
       "3         1          CaLo  09-25-14  Metabolites  Plasma  JCVI-00008   \n",
       "4         1          MaLo  09-25-14  Metabolites  Plasma  JCVI-00009   \n",
       "\n",
       "   Individual ID 2                               Sample ID 1  \\\n",
       "0                2  001_CaLo_04-16-14_Metabolites_JCVI-00001   \n",
       "1                2  001_CaLo_04-16-14_Metabolites_JCVI-00002   \n",
       "2                2  001_CaLo_07-11-14_Metabolites_JCVI-00003   \n",
       "3                2  001_CaLo_09-25-14_Metabolites_JCVI-00008   \n",
       "4                3  001_MaLo_09-25-14_Metabolites_JCVI-00009   \n",
       "\n",
       "                                      Sample ID 2  Unnamed: 9 Individual Key  \\\n",
       "0  001_002_Metabolites_Plasma_JCVI-00001_04-16-14         NaN           TaLo   \n",
       "1  001_002_Metabolites_Plasma_JCVI-00002_04-16-14         NaN           CaLo   \n",
       "2  001_002_Metabolites_Plasma_JCVI-00003_07-11-14         NaN           MaLo   \n",
       "3  001_002_Metabolites_Plasma_JCVI-00008_09-25-14         NaN           GrLo   \n",
       "4  001_003_Metabolites_Plasma_JCVI-00009_09-25-14         NaN           ElGo   \n",
       "\n",
       "   Unnamed: 11  Family Key  Unnamed: 13  \n",
       "0            1  Longstreet            1  \n",
       "1            2         NaN          NaN  \n",
       "2            3         NaN          NaN  \n",
       "3            4         NaN          NaN  \n",
       "4            5         NaN          NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "samples = [sample for sample in os.listdir(microbiome_path)\n",
    "           if os.path.isdir(os.path.join(microbiome_path, sample)) and \n",
    "           sample not in ['Fastq', 'summary', 'new']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sample_map = dict((sample, manifest[manifest['VendorID'] == sample]['Sample ID 2'].values[0])\n",
    "                  for sample in samples)\n",
    "sort_by_date = lambda x: datetime.strptime(x[1].split('_')[-1], \"%m-%d-%y\")\n",
    "sorted_sample_map = sorted(sample_map.items(), key=sort_by_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "files = [fn for fn in os.listdir(os.path.join(microbiome_path, samples[1]))\n",
    "         if fn.endswith('.txt')]\n",
    "files = [fn for fn in files\n",
    "         if not ((fn.startswith('Virus') and (fn.split('-')[1] != 'toprank')))]\n",
    "files = [fn for fn in files\n",
    "         if not fn.startswith('Virus')]\n",
    "files = [fn for fn in files if fn not in ['protein-ann.txt']]\n",
    "files.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def create_df(fn):\n",
    "    def load_df(file_path, sample_id):\n",
    "        df = pd.read_table(file_path)\n",
    "        indices = set(df.columns)\n",
    "        file_type = os.path.basename(file_path).split('-')[0]\n",
    "        if file_type in ['Eukaryota', 'Bacteria', 'Viruses']:\n",
    "            indices = indices-set(['abundance'])\n",
    "            indices = list(indices)\n",
    "            df.set_index(indices, inplace=True, drop=True)\n",
    "        elif file_type in ['superkingdom']:\n",
    "            indices = indices-set(['reads_count'])\n",
    "            indices = list(indices)\n",
    "            df.set_index(indices, inplace=True, drop=True)\n",
    "        else:\n",
    "            indices = indices - set(['Coverage', 'No_orfs', 'Abundance'])\n",
    "            indices = list(indices)\n",
    "            df.set_index(indices, inplace=True, drop=True)\n",
    "            df = df[['Abundance']]\n",
    "        df.columns = [sample_id]\n",
    "        return df\n",
    "    dfs = []\n",
    "    for sample, sample_id in sorted_sample_map:\n",
    "        #print sample\n",
    "        file_path = os.path.join(microbiome_path, \n",
    "                                 sample,\n",
    "                                 fn)\n",
    "        dfs.append(load_df(file_path, sample_id))\n",
    "    return pd.concat(dfs, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bacteria-class-abundance.txt\n",
      "Bacteria-family-abundance.txt\n",
      "Bacteria-genus-abundance.txt\n",
      "Bacteria-order-abundance.txt\n",
      "Bacteria-phylum-abundance.txt\n",
      "Bacteria-species-abundance.txt\n",
      "Bacteria-toprank-abundance.txt\n",
      "Eukaryota-class-abundance.txt\n",
      "Eukaryota-family-abundance.txt\n",
      "Eukaryota-genus-abundance.txt\n",
      "Eukaryota-order-abundance.txt\n",
      "Eukaryota-phylum-abundance.txt\n",
      "Eukaryota-species-abundance.txt\n",
      "Eukaryota-toprank-abundance.txt\n",
      "cog-ann.txt\n",
      "cog-class-ann.txt\n",
      "kog-ann.txt\n",
      "kog-class-ann.txt\n",
      "pfam-ann.txt\n",
      "superkingdom-reads-count.txt\n",
      "tigrfam-ann.txt\n",
      "tigrfam-mainrole-ann.txt\n",
      "tigrfam-subrole-ann.txt\n"
     ]
    }
   ],
   "source": [
    "out_file = \"/home/kunal/Schork-11-samples-taxon-and-function.xlsx\"\n",
    "writer = pd.ExcelWriter(out_file)\n",
    "for fn in files:\n",
    "    print fn\n",
    "    df = create_df(fn).fillna(0)\n",
    "    df.to_csv(\"/home/kunal/tscc_projects/tanner/data/family3/microbiome/14009b/new/\"+fn,\n",
    "             sep=\"\\t\")\n",
    "    df.to_excel(writer, fn)\n",
    "writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
